File: Computer Programming/HW3/problem2/src/FrontEnd.java
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.*;
import java.time.LocalDateTime;

public class FrontEnd {
    private UserInterface ui;
    private BackEnd backend;
    private User user;

    public FrontEnd(UserInterface ui, BackEnd backend) {
        this.ui = ui;
        this.backend = backend;
    }
    
    public boolean auth(String authInfo){
        String id = authInfo.split("\n")[0];
        String passwd = authInfo.split("\n")[1];
        User user = backend.auth(id, passwd);
        if (user == null) {
            return false;
        } else {
            this.user = user;
            return true;
        }
    }

    public void post(Pair<String, String> titleContentPair) {
        String id = titleContentPair.key;
        String content = titleContentPair.value;
        backend.post(id, content, user);
    }
    
    public void recommend(int N){
        List<Post> friendPosts = backend.recommend(user);
        if (friendPosts != null) {
            if (friendPosts.size() < N) {
                for (Post post : friendPosts) {
                    ui.println(post);
                }
            } else {
                for (int i = 0; i < N; i++) {
                    ui.println(friendPosts.get(i));
                }
            }
        }
    }

    public void search(String command) {
        String[] inputs = command.split(" ");
        Set<String> keywords = new HashSet<>(Arrays.asList(inputs).subList(1, inputs.length));
        List<Post> posts = backend.search(keywords);
        if (posts != null) {
            if (posts.size() < 10) {
                for (Post post : posts) {
                    ui.println(post.getSummary());
                }
            } else {
                for (int i = 0; i < 10; i++) {
                    ui.println(posts.get(i).getSummary());
                }
            }
        }
    }
    
    User getUser(){
        return user;
    }
}

File: Computer Programming/HW6/problem2/frontend.cpp
#include <set>
#include "frontend.h"

FrontEnd::FrontEnd(BackEnd* backend) : backend(backend) { }

bool FrontEnd::auth_id(string id_input) {
    return backend->auth_id(id_input);
}

bool FrontEnd::auth(string id_input, string pw_input) {
    User* user = backend->auth(id_input, pw_input);
    if (user == nullptr) {
        return false;
    } else {
        this->user = user;
        return true;
    }
}

void FrontEnd::post(istream& is, ostream& os) {
    string title, content, contents;

    os << "-----------------------------------\nNew Post\n* Title=";
    getline(is, title);

    os << "* Content\n>";
    getline(is, content);
    contents += "\n" + content;
    os << ">";
    while (true) {
        getline(is, content);
        if (content.empty()) {
            break;
        }
        contents += "\n" + content;
        os << ">";
    }

    backend->post(user->id, title, contents);
}

void FrontEnd::recommend(std::istream&is, std::ostream& os, int i) {
    vector<Post> posts = backend->recommend(user->id, i);
    for (Post post : posts) {
        os << "-----------------------------------\n" <<
              "id: " << post.id << "\n" <<
              "created at: " << post.date << "\n" <<
              "title: " << post.title << "\n" <<
              "content:\n";
        for (string content : post.contents) {
            os << content << "\n";
        }
    }
}

void FrontEnd::search(std::istream& is, std::ostream& os, set<string> keywords) {
    set<Pair> posts = backend->search(keywords);
    if (!posts.empty()) {
        os << "-----------------------------------\n";
    }

    auto iter = posts.begin();
    for (int count = 0; count < 10; count++) {
        if (iter == posts.end()) {
            break;
        }
        os << "id: " << (*iter).post.id <<
              ", created at: " << (*iter).post.date <<
              ", title: " << (*iter).post.title << "\n";
        iter++;
    }
}

User* FrontEnd::get_user() {
    return user;
}